src/main/java/gregtech/common/tileentities/machines/basic/GT_MetaTileEntity_BasicDrillerBase.java

```diff
@@ -0,0 +1,138 @@
+import net.minecraft.item.ItemStack;
+import net.minecraft.nbt.NBTTagCompound;
+import net.minecraftforge.common.util.FakePlayer;
+import org.lwjgl.Sys;
+
+import static gregtech.api.enums.GT_Values.V;
+
+public abstract class GT_MetaTileEntity_BasicDrillerBase extends GT_MetaTileEntity_BasicMachine {
+
+    static int[] RADIUS, //Miner radius per tier
+    SPEED, //Miner cycle time per tier
+    ENERGY ; //Miner energy consumption per tier
+
+    protected static final ItemStack MINING_PIPE = GT_ModHandler.getIC2Item("miningPipe", 0);
+    protected static final Block MINING_PIPE_BLOCK = GT_Utility.getBlockFromStack(MINING_PIPE);
+    protected static final Block MINING_PIPE_TIP_BLOCK = GT_Utility.getBlockFromStack(GT_ModHandler.getIC2Item("miningPipeTip", 0));
+
+    int drillX, drillY, drillZ;
+    boolean isPickingPipes;
+    boolean waitMiningPipe = true;
+
+    public GT_MetaTileEntity_BasicDrillerBase(int aID, String aName, String aNameRegional, int aTier, String[] aDescription, int aInSlots, int aOutSlots, String aGUIName, ITexture... aTextures){
+        super(aID,aName,aNameRegional,aTier,1,aDescription,aInSlots,aOutSlots,aGUIName,"",aTextures);
+    }
+
+    public GT_MetaTileEntity_BasicDrillerBase(String aName, int aTier, String[] aDescription, ITexture[][][] aTextures, int aInputSlotCount, int aOutputSlotCount, String aGUIName) {
+        super(aName,aTier,1,aDescription,aTextures,aInputSlotCount,aOutputSlotCount,aGUIName,"");
+    }
+
+    public boolean allowPutStack(IGregTechTileEntity aBaseMetaTileEntity, int aIndex, byte aSide, ItemStack aStack) {
+        return (super.allowPutStack(aBaseMetaTileEntity, aIndex, aSide, aStack)) && (aStack.getItem() == MINING_PIPE.getItem());
+    }
+
+    public abstract boolean hasFreeSpace();
+
+    @Override
+    public void onPostTick(IGregTechTileEntity aBaseMetaTileEntity, long aTick) {
+        super.onPostTick(aBaseMetaTileEntity, aTick);
+        if (aBaseMetaTileEntity.isServerSide()) {
+            if (aBaseMetaTileEntity.isAllowedToWork() && aBaseMetaTileEntity.isUniversalEnergyStored(ENERGY[mTier] * (SPEED[mTier] - mProgresstime)) && hasFreeSpace()) {
+                miningPipe:
+                if (waitMiningPipe) {
+                    mMaxProgresstime = 0;
+                    for (int i = 0; i < mInputSlotCount; i++) {
+                        ItemStack s = getInputAt(i);
+                        if (s != null && s.getItem() == MINING_PIPE.getItem() && s.stackSize > 0) {
+                            waitMiningPipe = false;
+                            break miningPipe;
+                        }
+                    }
+                    return;
+                }
+                aBaseMetaTileEntity.decreaseStoredEnergyUnits(ENERGY[mTier], true);
+                mMaxProgresstime = SPEED[mTier];
+            } else {
+                mMaxProgresstime = 0;
+                return;
+            }
+            if (mProgresstime == SPEED[mTier] - 1) {
```

Comment:
Выражение SPEED[mTier] - 1 стоит тоже вынести, но его смысл для меня непонятен, поэтому на ваше усмотрение.
```suggestion
            if (mProgresstime != SPEED[mTier] - 1) {
                return;
            }
```

Reply:
это самый банальный счетчик тиков, если поставить =! то будет работать каждый тик за исключением заданного значения => неверная работа

Reply:
Обращаю внимание на 68 строку в ченже.